src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/DefaultHybridCache.StampedeStateT.cs

```diff
@@ -175,29 +176,64 @@ private async Task BackgroundFetchAsync()
+                    T newValue = await _underlying!(_state!, SharedToken).ConfigureAwait(false);
+
+                    // If we're writing this value *anywhere*, we're going to need to serialize; this is obvious
+                    // in the case of L2, but we also need it for L1, because MemoryCache might be enforcing
+                    // SizeLimit (we can't know - it is an abstraction), and for *that* we need to know the item size.
+                    // Likewise, if we're writing to a MutableCacheItem, we'll be serializing *anyway* for the payload.
+                    //
+                    // Rephrasing that: the only scenario in which we *do not* need to serialize is if:
+                    // - it is an ImmutableCacheItem
+                    // - we're writing neither to L1 nor L2
+
+                    const HybridCacheEntryFlags DisableL1AndL2 = HybridCacheEntryFlags.DisableLocalCacheWrite | HybridCacheEntryFlags.DisableDistributedCacheWrite;
+                    var cacheItem = CacheItem;
+                    bool skipSerialize = cacheItem is ImmutableCacheItem<T> && (Key.Flags & DisableL1AndL2) == DisableL1AndL2;
+
+                    if (skipSerialize)
                     {
-                        if (cacheItem.TryReserveBuffer(out var buffer))
-                        {
-                            // mutable: we've already serialized it for the shared cache item
-                            await Cache.SetL2Async(Key.Key, in buffer, _options, SharedToken).ConfigureAwait(false);
-                            _ = cacheItem.Release(); // because we reserved
-                        }
-                        else if (cacheItem.TryGetValue(out var value))
+                        SetImmutableResultWithoutSerialize(newValue);
+                    }
+                    else if (cacheItem.TryReserve())
+                    {
+                        // ^^^ The first thing we need to do is make sure we're not getting into a thread race over buffer disposal.
+                        // In particular, if this cache item is somehow so short-lived that the buffers would be released *before* we're
+                        // done writing them to L2, which happens *after* we've provided the value to consumers.
+                        var writer = RecyclableArrayBufferWriter<byte>.Create(MaximumPayloadBytes); // note this lifetime spans the SetL2Async
+                        var serializer = Cache.GetSerializer<T>();
+                        serializer.Serialize(newValue, writer);
+                        BufferChunk buffer = new(writer.DetachCommitted(out var length), length, returnToPool: true); // remove buffer ownership from the writer
+                        writer.Dispose(); // we're done with the writer
+
+                        // protect "buffer" (this is why we "reserved"); we don't want SetResult to nuke our local
+                        var snapshot = buffer;
```

Comment:
nit: Can we call `buffer.DoNotReturnToPool()` here so that `snapshot` and `buffer` don't both claim ownership of the pooled array? I think, from reading the code, that `buffer` isn't stored then released anywhere (currently), but it'd be nice to be explicit and avoid any future mistakes.

Reply:
Yes; done, along with extra comments and renamed `snapshot` to be more meaningful